# Global Register of Introduced and Invasive Species - Australia (GRIIS)
Import GRIIS list from [GBIF IPT](https://cloud.gbif.org/griis/resource?r=griis-australia)

In [4]:
#import essential libraries
import pandas as pd
import requests
import os
import zipfile
from pathlib import Path

Download the Darwin Core Archive and read the species list into a dataframe

In [14]:
url = "https://cloud.gbif.org/griis/archive.do?r=griis-australia&v=1.9"
basedir = "/Users/oco115/PycharmProjects/authoritative-lists/"
sourcedir = basedir+"source-data/griis/"
targetdir = basedir+"current-lists/griis/"
print("Downloading dwca")
r = requests.get(url)

Retrieve the filename from the headers. Download and unzip the file

In [15]:
#get the filename from the headers and create a directory holder for it
basedir = "/Users/oco115/PycharmProjects/authoritative-lists/"
sourcedir = basedir+"source-data/griis/"
targetdir = basedir+"current-lists/griis/"
dwca = sourcedir + r.headers['content-disposition'].split("=")[1].replace('"','')
if Path(dwca).exists():
  os.remove(dwca)
unzipdirname = dwca[:-len(".zip")]
if not Path(unzipdirname):
  os.mkdir(unzipdirname)
#downlaad
with open(dwca, "wb") as output_file:
    output_file.write(r.content)
# unzip
with zipfile.ZipFile(dwca, 'r') as z:
   z.extractall(unzipdirname)

Explore the list, looking especially for encoding issues.

In [16]:
#explore
taxondf = pd.read_csv(unzipdirname + os.sep + "taxon.txt",sep="\t",lineterminator="\n")
taxondf

id  taxonID                                 scientificName  \
0     151000   151000  Abelia ×grandiflora (Rovelli ex André) Rehder   
1     151001   151001                Abelmoschus manihot (L.) Medik.   
2     151002   151002                    Abutilon grandiflorum G.Don   
3     151003   151003       Abutilon pictum (Gillies ex Hook.) Walp.   
4     151004   151004                    Abutilon theophrasti Medik.   
...      ...      ...                                            ...   
2947  153979   153979                          Ziziphus jujuba Mill.   
2948  153980   153980                       Ziziphus mauritiana Lam.   
2949  153981   153981                      Ziziphus mucronata Willd.   
2950  153982   153982              Ziziphus spina-christi (L.) Desf.   
2951  153983   153983  Zoobotryon verticillatum (Delle Chiaje, 1822)   

                                      acceptedNameUsage   kingdom  \
0                                                   NaN   Plantae   
1                                                   NaN   Plantae   
2                                                   NaN   Plantae   
3     Callianthe picta (Gillies ex Hook. & Arn.) Don...   Plantae   
4                                                   NaN   Plantae   
...                                                 ...       ...   
2947                                                NaN   Plantae   
2948                                                NaN   Plantae   
2949                                                NaN   Plantae   
2950                                                NaN   Plantae   
2951          Amathia verticillata (delle Chiaje, 1822)  Animalia   

            phylum          class           order          family taxonRank  \
0     Tracheophyta  Magnoliopsida      Dipsacales  Caprifoliaceae   SPECIES   
1     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
2     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
3     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
4     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
...            ...            ...             ...             ...       ...   
2947  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2948  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2949  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2950  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2951       Bryozoa   Gymnolaemata  Ctenostomatida  Vesiculariidae   SPECIES   

     taxonomicStatus  
0           ACCEPTED  
1           ACCEPTED  
2           ACCEPTED  
3            SYNONYM  
4           ACCEPTED  
...              ...  
2947        ACCEPTED  
2948        ACCEPTED  
2949        ACCEPTED  
2950        ACCEPTED  
2951         SYNONYM  

[2952 rows x 11 columns]

Problem records found by eyeballing. The file specifies UTF-8 encoding but has been encoded in something else.

In [17]:
taxondf.loc[taxondf['taxonID'].isin([151000,151155,151156,151617,151634,152066,153106])][['taxonID','scientificName']]

taxonID                                 scientificName
0      151000  Abelia ×grandiflora (Rovelli ex André) Rehder
154    151155        Amphilophus citrinellus (Günther, 1864)
155    151156           Amphilophus labiatus (Günther, 1864)
612    151617        Cichlasoma trimaculatum (Günther, 1867)
629    151634              Cladophora prolifera (Roth) Kütz.
1059   152066               Fallopia convolvulus (L.) Á.Löve
2080   153106                     Pinus durangensis Martínez

These values should be:

151000 Abelia x grandiflora (Rovelli ex André) Rehder
151155 Amphilophus citrinellus (Günther, 1864)
151156 Amphilophus labiatus (Günther, 1864)
153106 Pinus durangensis Martínez

There are many more though, just by searching for a single character:

In [18]:
taxondf.loc[taxondf['scientificName'].str.contains('Ã')]

id  taxonID                                 scientificName  \
855   151862   151862               Dianella ensifolia (L.) Redouté   
891   151898   151898            Dipolydora flava (Claparède, 1870)   
1636  152656   152656              Manihot carthaginensis Müll.Arg.   
2253  153281   153281                     Quercus palustris Münchh.   
2444  153472   153472  Schoenoplectus californicus (C.A.Mey.) Soják   

     acceptedNameUsage   kingdom        phylum          class         order  \
855                NaN   Plantae  Tracheophyta     Liliopsida   Asparagales   
891                NaN  Animalia      Annelida     Polychaeta      Spionida   
1636               NaN   Plantae  Tracheophyta  Magnoliopsida  Malpighiales   
2253               NaN   Plantae  Tracheophyta  Magnoliopsida       Fagales   
2444               NaN   Plantae  Tracheophyta     Liliopsida        Poales   

             family taxonRank taxonomicStatus  
855   Asphodelaceae   SPECIES        ACCEPTED  
891       Spionidae   SPECIES        ACCEPTED  
1636  Euphorbiaceae   SPECIES        ACCEPTED  
2253       Fagaceae   SPECIES        ACCEPTED  
2444     Cyperaceae   SPECIES        ACCEPTED

ftfy (fixed that for you) library works magic on these

In [19]:
import ftfy
from ftfy import fix_encoding
sample = "Abelia ×grandiflora (Rovelli ex André) Rehder"
fix_encoding(sample)

'Abelia ×grandiflora (Rovelli ex André) Rehder'

Apply to the `scientificName` and `acceptedNameUsage` fields

In [22]:
taxondf['scientificName'] = taxondf['scientificName'].apply(fix_encoding)
taxondf[['taxonID','scientificName','acceptedNameUsage']]

taxonID                                 scientificName  \
0      151000  Abelia ×grandiflora (Rovelli ex André) Rehder   
1      151001                Abelmoschus manihot (L.) Medik.   
2      151002                    Abutilon grandiflorum G.Don   
3      151003       Abutilon pictum (Gillies ex Hook.) Walp.   
4      151004                    Abutilon theophrasti Medik.   
...       ...                                            ...   
2947   153979                          Ziziphus jujuba Mill.   
2948   153980                       Ziziphus mauritiana Lam.   
2949   153981                      Ziziphus mucronata Willd.   
2950   153982              Ziziphus spina-christi (L.) Desf.   
2951   153983  Zoobotryon verticillatum (Delle Chiaje, 1822)   

                                      acceptedNameUsage  
0                                                   NaN  
1                                                   NaN  
2                                                   NaN  
3     Callianthe picta (Gillies ex Hook. & Arn.) Don...  
4                                                   NaN  
...                                                 ...  
2947                                                NaN  
2948                                                NaN  
2949                                                NaN  
2950                                                NaN  
2951          Amathia verticillata (delle Chiaje, 1822)  

[2952 rows x 3 columns]

Write the file and remove the original zip download.

In [24]:
taxondf.to_csv(targetdir+"griis-1.9.csv",index=False)
# os.remove(dwca)